Internship examples

agarciadiego opened the pull request:
## Fixes # .
None

## Proposed changes:
- Adds examples to generic property package

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

ksbeattie commented:
@agarciadiego, any update on this?

agarciadiego commented:
@ksbeattie I am trying to fix the error that is failing the tests, however tbh I do not know what the error is

ksbeattie commented:
@agarciadiego, I just updated from main and all but the circleci tests passed.  @lbianchi-lbl, I think we don't need/want the CircleCI test anymore, is that right?

Eitherway, @agarciadiego, if you are done making changes to this PR, I think you should remove the "draft" status and chase people down for reviews.

andrewlee94 commented on src/Examples/Advanced/CustomProperties/Hydrocarbon_Processing_example.ipynb:
```diff
@@ -0,0 +1,1078 @@
+   "source": [
+    "## 1. Introduction\n",
+    "\n",
+    "This Jupyter Notebook solves and optimizes a simple hydrocarbon processing unit. A property package has been created for hydrocarbon processing wivh includes methane, ethane, propane, n-butane, i-butane, ethylene, propene, butene, pentene, hexene, heptene and octene. Additionally, a sensitivity study is performed over the influence of temperature in the Gibbs reactor.\n",
```
Typo: which

andrewlee94 commented on src/Examples/Advanced/CustomProperties/Hydrocarbon_Processing_example.ipynb:
```diff
@@ -0,0 +1,1078 @@
+    "* Utilize the heat exchanger, gibbs reactor and flash unit models.\n",
+    "* Utilize the Generic Property Package, which provides a flexible platform for users to build property packages by calling upon libraries of modular sub-models to build up complex property calculations with the least effort possible.\n",
+    "* Demonstrate the use of arcs as streams connecting unit models\n",
+    "* Save a jazon file to avoid initialization if it has been done before.\n",
```
Typo: json

andrewlee94 commented on src/Examples/Advanced/CustomProperties/Hydrocarbon_Processing_example.ipynb:
```diff
@@ -0,0 +1,1078 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "For this specific problem we will need two different property packages as we will utilize an only vapor property package for the Gibbs reactor and a VLE package for the flash column"
```
vapor-only would probably read better here.

andrewlee94 commented on src/Examples/Advanced/CustomProperties/Hydrocarbon_Processing_example.ipynb:
```diff
@@ -0,0 +1,1078 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 3.4 Results"
```
A little more explanation of the results would probably be useful so the reader understands what is happening and why.

andrewlee94 commented on src/Examples/Advanced/CustomProperties/Hydrocarbon_Processing_example.ipynb:
```diff
@@ -0,0 +1,1078 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Hydrocarbon Processing with cubic equation of state case study"
```
One general comment on this example - it takes a really long time to run for me. I think we need to consider how we are going to handle this, as my guess is that it would seriously impact the run time of our integration tests.

agarciadiego replied:
I will include the json file with the initialization to avoid the time it takes. I will leave the sensitivity analysis but I will comment it out and add the plots as images so people can see how I do this but avoid the run time.

andrewlee94 commented on src/Examples/Advanced/CustomProperties/Dictionary_Txy_diagrams.ipynb:
```diff
@@ -0,0 +1,339 @@
+   "source": [
+    "## 1. Introduction\n",
+    "\n",
+    "This Jupyter Notebook creates a property package for the generic property package and utilized the Txy diagrams for the generic properties. A dictionary with the neccesary parameters.\n",
```
The second sentence here seems incomplete.

Also, I think a little more description of what this example shows and what the user will achieve by the end would be useful.

andrewlee94 commented on src/Examples/Advanced/CustomProperties/Parameter_estimation_PR.ipynb:
```diff
@@ -0,0 +1,950 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 3.1 Solving square problem"
```
This example fails for me at this point, with an infeasible solution. This might be due to changes in how scaling is implemented from recent PRs, although I tried quickly implementing the scaling tools which failed to fix the problem.

agarciadiego replied:
I will change the initialization method to initialize each point at the experimental composition. Also, I believe I may be adding a problem with the new scaling by scaling the objective function.

andrewlee94 reviewed (changes requested):
I tried running the three Jupyter notebooks, and got the following results:

1. The Txy diagram notebook worked.
2. The hydrocarbon example was extremely slow, and I ended up canceling the run after waiting a long time.
3. The ionic liquid parameter estimation failed to solve for the square problem with an infeasible solution.

I also found a few typos in the text that should be fixed.

andrewlee94 commented:
@agarciadiego Have you had a chance to look at the issues I encountered?

agarciadiego commented:
@andrewlee94 I have changed most of the typos and added explanations. I added a little bit of context on avoiding the errors and long run times in your comments. I would appreciate your input, especially on run times (how much time is too much?). I plan on committing all of this next week.

andrewlee94 commented:
@agarciadiego "How much time is too much" is difficult to give a specific answer to, but if it takes too long a user will either assume their notebook froze or just get fed up waiting. I would probably aim for only a couple of minutes at most if possible.

A few things to keep in mind:

1. Our automatic testing has a timeout (maybe 6 minutes from memory), so 5 minutes is probably a good guide as an upper bound.
2. From experience, it seems that long run times can cause issues with notebooks timing out internally. It still seems to work, but the output gets messed up.
3. If a cell is going to take more than a minute to run with no output, warn the user in the text beforehand so that they don't think their notebook froze.
4. If possible, give the user an easy way to control the number of data points/runs used. Start with a minimum number required to demonstrate the tool (ideally less than 2 minutes run time?), but show the user how they can add more data points if they want to run this themselves.

ksbeattie commented:
@agarciadiego, any news?

agarciadiego commented:
@andrewlee94 I applied the comments you made. I introduced a toggle function for the sensitivity analysis if people are interested in running it while making it known it will take a longer time. I also changed the initialization for the parameter estimation for each data point.

andrewlee94 commented on src/Examples/Advanced/CustomProperties/Dictionary_Txy_diagrams.ipynb:
```diff
@@ -216,7 +218,15 @@
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "2021-05-14 15:33:15 [INFO] idaes.generic_models.properties.core.generic.generic_property: DEPRECATED - definiton of generic property packages is moving to using static classes instead of modules. Please refer to the IDAES documentation.\n"
```
We should fix the issue that is causing this warning - it means that the config dictionary for the properties is using a module name to set some options, rather than a class in the module.

andrewlee94 commented on src/Examples/Advanced/CustomProperties/Dictionary_Txy_diagrams.ipynb:
```diff
@@ -259,7 +269,11 @@
      "output_type": "stream",
      "text": [
-      "2021-03-05 10:08:54 [INFO] idaes.init.props: Property package initialization: optimal - Optimal Solution Found.\n",
+      "2021-05-14 15:33:16 [WARNING] idaes.core.util.scaling: Accessing missing scaling factor for props[1].mole_frac_phase_comp\n",
```
Not particularly important, but it might be nice to get rid of these, or at least inform the user that they will see them and that they should see documentation/examples of the scaling tools to know more.

andrewlee94 commented on src/Examples/Advanced/CustomProperties/Hydrocarbon_Processing_example.ipynb:
```diff
@@ -151,6 +151,7 @@
      "name": "stdout",
      "output_type": "stream",
      "text": [
+      "2021-05-14 15:17:11 [INFO] idaes.generic_models.properties.core.generic.generic_property: DEPRECATED - definiton of generic property packages is moving to using static classes instead of modules. Please refer to the IDAES documentation.\n",
```
As before - lets try to fix the underlying issue (if they use the same property package, fixing one will fix both).

andrewlee94 reviewed (commented):
A couple more minor comments to try to clean up some deprecation warnings.

ksbeattie commented:
@agarciadiego, any news on this?

ksbeattie commented:
@MAZamarripa will be having a meeting with @agarciadiego on this, this week.

andrewlee94 commented on src/Examples/Advanced/CustomProperties/Parameter_estimation_PR.ipynb:
```diff
@@ -191,14 +191,13 @@
+      "2021-07-28 17:30:44 [INFO] idaes.init.fs.state_block: State variable initialization completed.\n",
+      "2021-07-28 17:30:44 [INFO] idaes.init.fs.state_block: Phase equilibrium initialization: infeasible - Converged to a locally infeasible point. Problem may be infeasible..\n",
+      "2021-07-28 17:30:44 [INFO] idaes.init.fs.state_block: Property initialization: infeasible - Converged to a locally infeasible point. Problem may be infeasible..\n",
+      "2021-07-28 17:30:44 [INFO] idaes.init.fs.state_block: Property package initialization: infeasible - Converged to a locally infeasible point. Problem may be infeasible..\n"
```
Just to check - I see an infeasible solution here. Is this a problem for the notebook/example?

If it isn't an issue, do we have a note to tell the reader not to worry and why?

agarciadiego replied:
I changed the initial conditions to avoid any confusion or the need to add a note.

andrewlee94 reviewed (commented):
Only one comment on an infeasible result I saw in the output.

andrewlee94 reviewed (approved):
I am told the test failure is a known issue and is not related to this PR. A fix is apparently in process.

lbianchi-lbl commented:
As a test, #57 is currently "successfully failing" in the same way (https://github.com/IDAES/examples-pse/pull/57/checks?check_run_id=3223757374), which is consistent with the Windows/Python 3.8 failing builds being unrelated to this PR.

MAZamarripa reviewed (approved):
lgtm
